Add terminal links to the Archivematica workflow

jorikvankemenade opened the pull request:
This PR includes three commits. We might want to squash these into a single commit before merging, but I think that at this point it makes sense to separate them. 

2e1f000 addresses the comment I made [here](https://github.com/archivematica/Issues/issues/1055#issuecomment-589644118). My jobs were not scheduled correctly because the `queues` class missed one of the links that indicate that a package is done. This sparked the discussion in [#1055](https://github.com/archivematica/Issues/issues/1055) about replacing the `PACKAGE_COMPLETED_LINK_IDS` altogether. This also removes `27fd07e-d3ed-4767-96a5-44a2251c6d0a` as a terminal link. Having this link marked as terminal causes some of the problems as discussed in [#1010](https://github.com/archivematica/Issues/issues/1010). 

efcc259 introduces the discussed solution. This commit basically adds an `is_terminal` function to each link in the workflow. This checks for the `end: true` value that can be added to the `workflow.json`.

Finally, bb6d3b4 applies this concept to the currently existing queueing logic. I have removed all the `PACKAGE_COMPLETED_LINK_IDS` and added the `end` field to each of these links.

I have tested the changes by running a few transfers, including some failing ones. I set `ARCHIVEMATICA_MCPSERVER_CONCURRENT_PACKAGES` to 1 to see if I could get the queueing of packages to block. So far I haven't, but maybe you guys have a more efficient way to test this.

Connects to https://github.com/archivematica/Issues/issues/1108.

jorikvankemenade committed:
queues: update PACKAGE_COMPLETED_LINK_IDS

Add bbfbecde-370c-4e26-8087-cfa751e72e6a (Failed Compliance) as a
terminal link. If we don't we can block the system if too many packages
fail.

jorikvankemenade committed:
workflow: add is_terminal check

Add an is_terminal property to the workflow. With this property it is
possible to indicate for every link if it is the end of a workflow. This
can be used by the queueing mechanism to see if the next job in the
chain needs to be scheduled, if a package needs to be (de)activated, or
if a new job needs to be queued from one of the existing queues.

Using this property is as simple as adding the boolean field "end" to
any link in the existing workflow. Using the property is as simple as
calling is_terminal() on any Link. This will retrun the value of the
"end" field, or false if the field does not exist.

jorikvankemenade committed:
queues: use workflow to indicate package ending

Set the "end" property on workflow links that indicate the end of the
processing of a package. Use this property to decide if a package needs
to be deactivated. This allows for manipulating the queueing and package
behaviour from the workflow, instead of hardcoded in the queueing logic.

In the future we might want to switch the is_terminal property to a less
explicit check. Instead of checking for the "end" property it should
check if there are no links to future jobs. If this is the case this
would signal the end of the chain. See issue #1055 for the full
discussion on this topic.

This commit also clarifies the behaviour of the package and job complete
callbacks. It specifically mentions the role of the callbacks with
respect to the job chains.

cole commented on src/MCPServer/lib/server/workflow.py:
```diff
@@ -194,6 +194,17 @@ def get_status_id(self, code):
             status_id = self._src["fallback_job_status"]
         return status_id
 
+    def is_terminal(self):
```
Nit: I'd use a property here and return `bool(self._src["end"])`. But it doesn't matter much.

jorikvankemenade replied:
What do you mean exactly? If I use something like this:

```python
    @property
    def is_terminal(self):
        """Check if the link is indicated as a terminal link.

        Returns the value of the end property.
        If the value does not exist the link is assumed to be non-terminal.
        """
        return self._src["end"]
```

It won't work, because `end` is not a required field, so you will get key errors. I can annotate the entire function with property, but what would be the advantage of that? I am not very familiar with the property concept.

sevein replied:
I think it's just a common idiom for getters in Python? But you're right that you'd still need the exception handler, unless you want to perform the computation in the constructor.

jorikvankemenade replied:
> I think it's just a common idiom for getters in Python?

I believe so too, it should return a read-only version of the value so that is nice. I have changed it to a property and replaced the `KeyError` part with a `get` method on the dict.

cole replied:
I meant:

```python
    @property
    def is_terminal(self):
        try:
            return bool(self._src["end"])
        except KeyError:
            return False
```

It sounds like you have this, which is more concise:
```python
    @property
    def is_terminal(self):
        return bool(self._src.get("end"))
```

This is a nitpick and a purely stylistic change. The original approach is also fine! In fact I think there's an argument to be made that it's more consistent with the rest of the module.

Generally I like `@property` for methods without arguments that return a single value and don't have external side effects. Calling `bool` ensures the method always returns a boolean; that's more a habit from using [type hints](https://docs.python.org/3/library/typing.html) than anything else; there's not a huge amount of benefit here.

cole reviewed (approved):
Much better! 🎉

sevein commented on src/MCPServer/lib/server/queues.py:
```diff
@@ -221,34 +197,33 @@ def stop(self):
-
-        # Special case for decision job here.
-        if isinstance(next_job, DecisionJob) and next_job.awaiting_decision:
+        elif isinstance(next_job, DecisionJob) and next_job.awaiting_decision:
```
I think that the previous code style was calculated. The early returns gets that falsy job case out of your way to let the function body continue dealing with other cases. There is not right or wrong here though, but `elif` implies previous context that we don't need anymore?

jorikvankemenade replied:
I see what you mean. The reason I changed it was because to me it all belongs to the same context, I see it as a switch on `next_job`. If there is no `next_job` we do nothing. In case we have a decision job we await the decision. In all other cases, we schedule the next job. 

The separation made me wonder why the early return was a special case. To me, it makes more sense now but if you prefer the old way I am more than happy to change it back.

sevein commented on src/MCPServer/lib/assets/workflow-schema-v1.1.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "$id": "https://www.archivematica.org/labs/workflow/schema/v1.json",
+  "$id": "https://www.archivematica.org/labs/workflow/schema/v1.1.json",
```
I would argue that we shouldn't remove the previous version. But let's just keep calling it v1 for now and avoid all the extra work around versioning until we promote this as a consumable language.

jorikvankemenade replied:
Yeah, I wasn't sure if versioning it was necessary or wanted. I'll revert the versioning changes.

sevein commented on src/MCPServer/lib/server/queues.py:
```diff
@@ -221,34 +197,33 @@ def stop(self):
+        It is assumed that a package is only complete when a terminal link is
+        hit. When we hit a terminal link the future should not contain the next
+        job in the chain. If the future is a new job we raise an exception.
         """
```
I like what you're doing here but I thought the bits about the execution context were also useful (*Called by an Executor...*). Didn't you find that useful when you read the method for the first time?

> If the future is a new job we raise an exception.

You've provided some useful context but I think we also have to draw a line where we stop commenting on *what* the code is doing because the code is self-describing. This idea I'm trying to communicate is better described [here](https://google.github.io/eng-practices/review/reviewer/looking-for.html#comments).

jorikvankemenade replied:
>  I thought the bits about the execution context were also useful

Originally I saw this as commenting about an implementation detail of the Python futures. However, looking at it again I see that this detail is actually coupled to how you can use the function. I.e. you cannot use the function to complete a package without using futures etc. So I agree that this adds some additional context that is relevant to the developer.

> This idea I'm trying to communicate is better described here.

I fully agree, missed the mark there. It is trivial to see in the code that we do this, and we already explained why - so that part was definitely obsolete.

sevein reviewed (changes requested):
~`"bbfbecde-370c-4e26-8087-cfa751e72e6a"` is now a terminal link but `"d27fd07e-d3ed-4767-96a5-44a2251c6d0a"` and `"032cdc54-0b9b-4caf-86e8-10d63efbaec0"` aren't. Do you mind explaning why?~

Sorry, I realise you've already explained it, thanks!

jorikvankemenade committed:
workflow: apply review comments

jorikvankemenade committed:
queues: apply review comments

jorikvankemenade commented:
@sevein I have added the requested changes in two new commits. If they are okay with you it is probably best to squash the whole PR into a single commit? Or I can squash the two new commits in the first three to keep some separation. Whatever you prefer.

sevein reviewed (changes requested):
@jorikvankemenade, I think I've found a couple of extra links that should terminate. Could you please confirm?

- `16415d2f` (Create SIPs from TRIM transfer containers) which is similar to "Move transfer to backlog" or "Create SIP from transfer objects" but when dealing with TRIM transfers,
- `f8ee488b` (Handle unstored DIP) should also be added? Otherwise the processing of DIPs won't finalize if the user chooses not to store DIPs.

On the other hand, I don't understand why you chose `bbfbecde`. Wouldn't make more sense to add `89071669` which sits right behind `/activeTranfsers/standardTransfer`?

I've updated amflow so it highlights terminal links in orange. Check it out if that makes things easier for you.

sevein commented on src/MCPServer/lib/server/queues.py:
```diff
@@ -221,34 +197,35 @@ def stop(self):
     def _package_completed_callback(self, package, future):
-        """Called by an Executor on completion the last job in a package.
+        """Called by an executor on completion of a package. Marks the package
+        as inactive and schedules a new package.
```
One nit about your docstrings in this PR:

https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings
> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description. The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line. The summary line may be on the same line as the opening quotes or on the next line. The entire docstring is indented the same as the quotes at its first line (see example below).

jorikvankemenade replied:
Check, this was a bit careless of me. Fixed it!

jorikvankemenade committed:
queues: fix docstring comments

jorikvankemenade commented:
>  I think I've found a couple of extra links that should terminate

I would imagine there can be a couple of them. I just refactored the existing known links and added the one I was missing in my workflow: standard transfers. So I think odds are more will be found when testing Archivematica, especially with `concurrent_packages = 1`. The advantage of this approach is that rectifying these bugs should be simple.

> I've updated amflow so it highlights terminal links in orange. 

This is actually pretty helpful. One thing which baffles me is that I have 11 terminal links in the file in this PR. And only 7 show up in the amflow. Are certain error paths not drawn?

> `16415d2f` (Create SIPs from TRIM transfer containers) which is similar to "Move transfer to backlog" or "Create SIP from transfer objects" but when dealing with TRIM transfers,

I agree that this is a similar situation.

> `f8ee488b` (Handle unstored DIP) should also be added? Otherwise the processing of DIPs won't finalize if the user chooses not to store DIPs.

This is the end of the workflow for DIP, so yes it is terminal. Actually, `653b134f` should be terminal as well. Another thing that I notice when checking the DIP section is that there is no failure path for DIPS. Is this on purpose? And if so, we should still make sure that we can deactivate the package on failure.

> On the other hand, I don't understand why you chose `bbfbecde`. Wouldn't it make more sense to add `89071669` which sits right behind `/activeTranfsers/standardTransfer`?

I see why you would think this. I just used the `workflow.json` to select this transfer. Because of the watched directory structure, I missed that this was not the end of a workflow.  Looking at the amflow I think that your suggestion makes more sense. 

One general concern that dawned on me when looking at this is that for many links there is no error path. I think this means that there are some points left where the workflow of a package can stop, without the package being deactivated. I think we have three directions here:
1. Role with the punches. Set all test machines to `concurrent_packages = 1` and add terminal links each time a failure is encountered.
1. Manually verify the entire chain and set failure states for all points in the workflow. This is a tedious error-prone job, but should result in a more reliable workflow.
1. Think of a fallback mechanism that deactivates packages on failure. 

What do you think @sevein? What is your preference?

jorikvankemenade committed:
workflow: add review comments

* Add 16415d2f because at this point a SIP is created for TRIM
  transfers.
* Add f8ee488b and 653b134f, these signal the end of the DIP packages.
* Exchange bbfbecde for 89071669, the package is done after the error
  path is completely finished. This is as soon as the transfer has
  reached the "normal" flow again.

sevein commented:
> This is actually pretty helpful. One thing which baffles me is that I have 11 terminal links in the file in this PR. And only 7 show up in the amflow. Are certain error paths not drawn?

amflow has to ignore some very common links because otherwise graphviz takes forever to render the graph (30s+). I should add an option to force the rendering of all links for cases like this, will look into that. This is the [list](https://github.com/artefactual-labs/amflow/blob/0113346fc47ddd911e433696e9aeec40bd2c9f5d/internal/graph/dotviz.go#L17-L29).

> This is the end of the workflow for DIP, so yes it is terminal. Actually, 653b134f should be terminal as well. Another thing that I notice when checking the DIP section is that there is no failure path for DIPS. Is this on purpose? And if so, we should still make sure that we can deactivate the package on failure.

You're right. I'll take a look.

> What do you think @sevein? What is your preference?

It's relatively easy to introduce some sanity checks in the existing amflow validation function which uses an actual graph internally. That graph does not omit links, it's only the graphviz encoder that ignores them.

I think I'll merge this now since it's become a blocker for us! As long as I have more feedback on missing links I'll submit a PR and will CC you! Thank you so much for the help, Jorik.

sevein reviewed (approved):
👍
